sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -105,6 +105,41 @@ abstract class Expression extends TreeNode[Expression] {
       val isNull = ctx.freshName("isNull")
       val value = ctx.freshName("value")
       val ve = doGenCode(ctx, ExprCode("", isNull, value))
+
+      // TODO: support whole stage codegen too
+      if (ve.code.trim.length > 1024 && ctx.INPUT_ROW != null && ctx.currentVars == null) {
+        val setIsNull = if (ve.isNull != "false" && ve.isNull != "true") {
+          val globalIsNull = ctx.freshName("globalIsNull")
+          ctx.addMutableState("boolean", globalIsNull, s"$globalIsNull = false;")
+          val localIsNull = ve.isNull
+          ve.isNull = globalIsNull
+          s"$globalIsNull = $localIsNull;"
+        } else {
+          ""
+        }
+
+        val setValue = {
+          val globalValue = ctx.freshName("globalValue")
+          ctx.addMutableState(
+            ctx.javaType(dataType), globalValue, s"$globalValue = ${ctx.defaultValue(dataType)};")
+          val localValue = ve.value
+          ve.value = globalValue
+          s"$globalValue = $localValue;"
+        }
+
+        val funcName = ctx.freshName(nodeName)
+        val funcFullName = ctx.addNewFunction(funcName,
+          s"""
+             |private void $funcName(InternalRow ${ctx.INPUT_ROW}) {
+             |  ${ve.code.trim}
+             |  $setValue
```

Comment:
Can we always pass `value` as a return value? It can reduce # of global variables.

Reply:
good suggestion! Actually, this is a general strategy which can be applied to more places. If there are only boolean global variables, it's very easy to fold them into one array.

Reply:
Thanks.

IMHO, I am curious whether we will see no performance degradation by using one array to compact many boolean variables.  
I am waiting for the updated result in [this discussion](https://github.com/apache/spark/pull/19518#issuecomment-337965330). This is because the current code seems to measure performance of interpreter due to lack of warmup.

Reply:
Is that a bad idea to prepare some utility classes to store a pair (value, isNull) for this splitting cases? I feel class fields are valuable resources.

Reply:
creating objects will be a big overhead. I think having a global boolean variable is better.

Reply:
I originally thought we could avoid the overhead by using thread-local singleton? But, it's a bit weird, so the current code looks good.

Reply:
The current code can pass the `value` as a local variable if this method is inlined, or can pass the `value` on register for return value if this method is not inlined.  
On the other hand, to use a object will always introduce memory accesses to access fields in caller and callee.